kが無限のとき、%o1の式を無限級数で表すと、次数の数を２つの３乗数の和で表す方法の個数がその次数の項の係数になります。

In [1]:
SN:sum(x^(n^3),n,1,k)^2;

                                   k
                                  ====    3
                                  \      n  2
(%o1)                            ( >    x  )
                                  /
                                  ====
                                  n = 1

kを有限にしても$(k^3+2)$次の項までの係数は正しく計算されます。k=12としてやってみましょう。下記の式では係数の不正確な$x^{k^3+2}$より次数の大きな項は捨てています。

In [2]:
TAXI:remainder(ev(SN,nouns),x^(k^3+2)),k:12;

         1729      1674      1547      1512    1458      1456      1395
(%o2) 4 x     + 2 x     + 2 x     + 2 x     + x     + 2 x     + 2 x
      1358      1343      1339      1332      1241      1216      1125
 + 2 x     + 2 x     + 2 x     + 2 x     + 2 x     + 2 x     + 2 x
      1072      1064      1027    1024      1008      1001      945      855
 + 2 x     + 2 x     + 2 x     + x     + 2 x     + 2 x     + 2 x    + 2 x
      854      793      756      737      730      728    686      637      576
 + 2 x    + 2 x    + 2 x    + 2 x    + 2 x    + 2 x    + x    + 2 x    + 2 x
      559      539      520      513      468    432      407      370      351
 + 2 x    + 2 x    + 2 x    + 2 x    + 2 x    + x    + 2 x    + 2 x    + 2 x
      344      341      280    250      243      224      217      189      152
 + 2 x    + 2 x    + 2 x    + x    + 2 x    + 2 x    + 2 x    + 2 x    + 2 x
      133    128      126      91      72      65    54      35      28    16
 + 2 x    + x    + 2 x

$x^{1729}$の係数が4になっていることは、$1729=12^3+1^3=1^3+12^3=10^3+9^3=9^3+10^3$と４通りに表すことができることに対応しています。
$x^{16}$の係数が１であることは、$16=2^3+2^3$と１通りに表すことができることに対応しています。

下記のコードで、上記の多項式で係数が4以上の項の次数を取り出すことが出来ます。

In [3]:
apply(append,map(lambda([term],
                    block([cf:coeff(term,x^hipow(term,x))],
                          if cf>3 then [hipow(term,x)] 
                          else [])),
                 args(TAXI)));

(%o3)                               [1729]

kの値を大きくすると%o1をそのまま計算するコードはどんどん遅くなります。それを避けるために、与えられたxの多項式polyの２乗を指定された次数maxdまで計算する関数

sqexpand(poly, maxd)

を下記のように定義します。

In [4]:
sqexpand(poly,maxd):=block([tlist:args(poly),res:0],
  for term in tlist do res:remainder(res+expand(term*poly),x^(maxd+1)),
  return(res));

(%o4) sqexpand(poly, maxd) := block([tlist : args(poly), res : 0], 
                                                               maxd + 1
for term in tlist do res : remainder(res + expand(term poly), x        ), 
return(res))

早速この関数を使ってk=450まで、次数としては$450^3+2$まで求めてみます。ただし出力項数が多すぎるため表示しません。

In [5]:
TT:sqexpand(sum(x^(n^3),n,1,450),450^3+2)$

結果を見るために、下記のコードを使い、係数が4以上の項の次数を出力します。また係数が５以上の項については係数と次数を出力します。

In [6]:
apply(append,map(lambda([term],
                    block([cf:coeff(term,x^hipow(term,x))],
                          if cf>4 then [["**",cf,"**",hipow(term,x)]]
                          elseif cf>3 then [hipow(term,x)]
                          else [])),
                 args(TT)));

(%o6) [91034307, 90527229, 90091008, 89576767, 88780536, 88129323, 88122125, 
87952501, 87699456, 87579037, [**, 6, **, 87539319], 87483968, 87029432, 
86960952, 86575168, 86488136, 86368464, 86316741, 86124824, 85622264, 
85502375, 85492792, 85188096, 84717568, 84637287, 84266000, 83758528, 
83393856, 81621568, 80779032, 80668224, 80294760, 80208576, 80102574, 
79692193, 79598125, 77414967, 77335776, 77129864, 76487168, 75550088, 
74130875, 74097261, 74093832, 73842713, 73819863, 73369088, 72669177, 
72131904, 71509312, 71292312, 70979896, 70957971, 70796808, 70449093, 
69934592, 69805125, 69625969, 69583995, 69177024, 68802048, 68696000, 
68007673, 67956616, 67944312, 67931136, 67173463, 66763333, 66469429, 
66201408, 65728000, 65573928, 65293317, 65272753, 65055744, 64623104, 
64421875, 64232000, 64125000, 63660032, 63057960, 62980288, 62135073, 
62108397, 61568208, 61555832, 61121024, 60965288, 60698521, 60271939, 
59913728, 58923072, 58751784, 58438072, 57066633, 56754152, 5673369

このリストの一番小さい数が１７２９であることが確認できます。また１７２９と同様に$450^3+2=91125002$までの数で、２つの３乗数で４通り以上の方法で表すことができる数が全てリストアップされています。さらに87539319については２つの３乗数の和で６通りに表すことができることがわかります。

Copyright 2021 Yasuaki Honda

### Computing Ramanujan's taxi number using the generating function $ (\sum_{n=1}^{\infty}　x^{n^3})^2 $